### DISCUSSION TITLE
Will the history plugin ever become a reality?

### DISCUSSION THREAD

**Discussion by hirasso** [alpinejs/alpine#3460] [4. General]
Hi there! A few days ago I noticed that history package that's linked from the readme: https://github.com/alpinejs/alpine/tree/main/packages/history

I think this would actually be a very nice thing to have! Right now, I need to do all that manually in my implementations, which is pretty painful especially if used together with `x-model`. So, any insight about the status of the history plugin would be much appreciated.

Have a great day everyone!

**Comment by ekwoka** [comment]
If you've got a painful working version yourself, you could try to make that PR to get something that works. I think it was stopped partially because of the challenge of getting it to really do what you expect it all to do.

**Comment by hirasso** [comment]
My version is very loosely coupled with Alpine.js and therefor not PR-material. I'm only using it for primitives, though (e.g. a searchstring and/or a filterstring). Maybe adding this constraint would be a way to make it more manageable? The `interceptor` function is [marked as private](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/interceptor.js#L1) right now, otherwise I would maybe try my luck.

**Reply by ekwoka** [reply]
It being private means it's fine to use in the package itself lol

Making it work with objects and arrays shouldn't be too tough. Just decide on a way to represent them (do you base64 URI encode JSON, or spread it to multiple nested params?) and do it same both in and out.

I believe in you!

**Reply by hirasso** [reply]
How would you combine a getter and setter with the `$queryString` magic?

**Comment by hirasso** [comment]
Actually... I just ran a quick test and it basically *just works* as it is right now, at least for primitive types! 😃

In the following component, both `count` as well as `search` are being updated nicely, even in combination with `x-model`. 

The only thing I noticed that doesn't quite feel right is that the query is not only updated after data mutation, but also when first visiting the page. I would prefer if it would just do nothing initially and wait for the first mutation of the data before doing it's thing.

```html
<div x-data="{ count: $queryString(5), search: $queryString('') }">
  <div x-text="count"></div>
  <button x-on:click="count--">-</button>
  <button x-on:click="count++">+</button>
  <hr>
  <input type="text" x-model="search"></input>
</div>
```
```js
// just my local import paths, directly from the compiled repo
import Alpine from '../../../alpine/packages/alpinejs/dist/module.esm.js';
import history from '../../../alpine/packages/history/dist/module.esm.js';
Alpine.plugin(history);
Alpine.start();
```
Anyways... before I start diving deeper: @calebporzio, if you read this: Would you like to share with us why you stopped working on this package and if you would be interested in finishing it?

**Comment by joshhanley** [comment]
@hirasso yeah, the history plugin is currently being worked on as Livewire V3 needs it. So I probably would hold off on doing anything to it as Caleb is currently working on it. You can see the latest changes here https://github.com/alpinejs/alpine/commit/3e6609663a96afb6c29974fdcca4fec6e1df97d4

**Reply by hirasso** [reply]
Ooof, that's such great news!! Thanks @joshhanley for saving me from submitting a hacky, clueless PR 😅 🙏

...also, a big thank you to all @alpinejs/maintainers for your quick and thoughtful answers. Much appreciated!

**Reply by joshhanley** [reply]
@hirasso no worries! 🙂
